agent/eni/watcher.go

```diff
@@ -0,0 +1,255 @@
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package eni
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/deniswernert/udev"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+
+	"github.com/aws/amazon-ecs-agent/agent/eni/netlinkWrapper"
+	"github.com/aws/amazon-ecs-agent/agent/eni/udevWrapper"
+	eniUtils "github.com/aws/amazon-ecs-agent/agent/eni/utils"
+)
+
+const (
+	udevSubsystem                 = "SUBSYSTEM"
+	udevNetSubsystem              = "net"
+	udevPCISubsystem              = "pci"
+	udevEventAction               = "ACTION"
+	udevAddEvent                  = "add"
+	udevRemoveEvent               = "remove"
+	udevDevPath                   = "DEVPATH"
+	udevInterface                 = "INTERFACE"
+	defaultReconciliationInterval = time.Second * 30
+	mapCapacityHint               = 10
+)
+
+// Watcher exposes a method to check if MACAddress is present in observed state
+type Watcher interface {
+	// IsMACAddressPresent returns a boolean to determine if MACAddress is present in observed state
+	IsMACAddressPresent(mac string) bool
+}
+
+// UdevWatcher maintains the state of attached ENIs
+// to the instance. It also has supporting elements to
+// maintain consistency and update intervals
+type UdevWatcher struct {
+	ctx                  context.Context
+	cancel               context.CancelFunc
+	updateLock           sync.RWMutex
+	updateIntervalTicker *time.Ticker
+	enis                 map[string]string // MAC => Device-Name
+	netlinkClient        netlinkWrapper.NetLink
+	udevMonitor          udevWrapper.Udev
+	events               chan *udev.UEvent
+}
+
+// New is used to return an instance of the UdevWatcher struct
+func New(ctx context.Context, nlWrap netlinkWrapper.NetLink, udevWrap udevWrapper.Udev) *UdevWatcher {
+	derivedContext, cancel := context.WithCancel(ctx)
+	return &UdevWatcher{
+		ctx:           derivedContext,
+		cancel:        cancel,
+		enis:          make(map[string]string, mapCapacityHint),
+		netlinkClient: nlWrap,
+		udevMonitor:   udevWrap,
+		events:        make(chan *udev.UEvent),
+	}
+}
+
+// Init initializes a new ENI Watcher
+func (udevWatcher *UdevWatcher) Init() error {
+	links, err := udevWatcher.netlinkClient.LinkList()
+	if err != nil {
+		return errors.Wrapf(err, "init udevWatcher: error retrieving network interfaces")
+	}
+
+	udevWatcher.updateLock.Lock()
+	for _, link := range links {
+		deviceName := link.Attrs().Name
+		macAddress := link.Attrs().HardwareAddr.String()
+		if macAddress != "" {
+			udevWatcher.addDeviceWithMACAddressUnsafe(deviceName, macAddress)
+		}
+	}
+	udevWatcher.updateLock.Unlock()
+
+	// UDev Event Handler
+	go udevWatcher.eventHandler(udevWatcher.ctx)
```

Comment:
My understanding is that the "initialize" is only responsible for preparing the resource, and the actually action should be triggered in "Start()", in this way the initialize will start the process of listening to the event and updating the map?